File: shuffle/ans.pl
#!perl

use strict;
use warnings;

my $f = shift;
open my $fh, '<', $f or die "cannot open file : $!";

## 問題数
my $T = <$fh>;

my $cnt = 1;
while( my $line = <$fh> ) {
    chomp($line);

    # M, C, W
    my ($M, $C, $W) = split(/ /, $line);

    # cards
    my @cards = (1..$M);
    
    # swap
    for(my $i=1; $i<=$C; $i++) {
        $line = <$fh>;
        chomp($line);
        my ($A, $B) = split(/ /, $line);

        my @part = splice(@cards, $A-1, $B);
        push(@part, @cards);
        @cards = @part;
    }
    
    # answer
    print "Case #$cnt: " . $cards[$W-1] . "\n";

    $cnt++;
}

File: shuffle/ans.rb
#! /usr/bin/ruby
# -*- coding: utf-8 -*-

f = ARGV[0]
fh = open(f, "r")

## problems
t = fh.gets

cnt = 1
while line = fh.gets
  # M, C, W
  m,c,w = line.split(" ")

  # cuts
  cuts = Array.new
  i = 0
  while i < c.to_i
    line = fh.gets
    a, b = line.split(" ")
    
    h = Hash.new
    h['A'] = a.to_i
    h['B'] = b.to_i
    cuts.push(h)

    i = i + 1
  end

  # reverse cut
  idx = w.to_i - 1
  cuts.reverse_each do |h|
    x = h['A'] - 1
    y = h['B']
    if idx < y
      idx = idx + x
    elsif idx < y + x
      idx = idx - y
    end
  end

  # answer
  print 'Case #', cnt, ': ', idx+1, "\n"

  cnt = cnt + 1
end
